base/ltpara.dtx

```diff
@@ -0,0 +1,1147 @@
+\cs_set_eq:NN \SuppressIndentationBox \para_suppress_indentation:
+%    \end{macrocode}
+%  \end{macro}
+%
+%
+%
+%
+%
+%
+%  \begin{macro}{\para_end:}
+%    Adding hooks to the end of a paragraph is similar but here we
+%    need to alter the command that is used by \TeX{} to end horizontal
+%    mode and return to vertical mode, i.e., \cs{par}.
+%
+%    This is a bit more complicated as this command can appear anywhere
+%    either explicitly or implicitly added by \TeX{} in certain
+%    situations:
+%    \begin{itemize}
+%    \item
+%      when using \cs{par} in the code or the document
+%    \item
+%      when using a blank line (which is converted to \cs{par})
+%    \item
+%      when \TeX{} finds any commands incompatible with horizontal
+%      mode it issues a \cs{par} and then rereads the command.
+%    \end{itemize}
+%
+%    Unfortunately, \TeX{} has some (these days) unnecessary
+%    optimization: if a \cs{vbox} ends and \TeX{} is still in
+%    horizontal mode it simply exercises the paragraph builder instead
+%    of issuing a \cs{par}. It is therefore necessary for \LaTeX{} to
+%    ensure that this case doesn't happen and all boxes internally
+%    have a \cs{par} command at their end.
+%
+%    This \cs{par} may or may not run the ``par primitive'' (which is
+%    always available as \cs{tex_par:D} in \pkg{expl3}); it is
+%    permissible to have a changed meaning and it is in fact changed
+%    by \LaTeX{} in various ways at various points inside
+%    \texttt{latex.ltx}.  For this \LaTeXe{} code has the following
+%    conventions: \cs{@@@@par} and \cs{endgraf} both refer to the
+%    default meaning (in the past this was the initex primitive) while
+%    \cs{par} is the current meaning which may does something else.
+%
+%
+%    We are now going to change this default meaning to run
+%    \cs{para_end:} instead, which ultimately executes the initex
+%    primitive but additionally adds our hooks when appropriate.
+%    This way the change is again transparent to the legacy \LaTeXe{}
+%    code.
+%
+%    In most cases \cs{para_end:} should behave exactly like the
+%    primitive and we achieve this by simply expanding it to the
+%    primitive which is available to us as \cs{tex_par:D}. This way we
+%    don't have to care about whether \TeX{} just does nothing (e.g.,
+%    if in vertical mode already) or generate an error, etc.
+%    \begin{macrocode}
+\cs_new_protected:Npn \para_end: {
+%    \end{macrocode}
+%
+%    The only case we care about is when we are in horizontal mode
+%    (i.e., doing typesetting) and not also in inner mode (i.e.,
+%    making paragraphs and not building an \cs{hbox}.
+%    \begin{macrocode}
+%  \bool_lazy_and:nnT
+%       { \mode_if_horizontal_p: }
+%       { \bool_not_p:n { \mode_if_inner_p: } }
+%       { ...
+%    \end{macrocode}
+%    This is inner loop so we code the above test with low-level \cs{if_mode...}s.
+%    \begin{macrocode}
+  \if_mode_horizontal:
+    \if_mode_inner: \else:
```

Comment:
I think these two conditionals should be replaced by the slightly higher level ones `\mode_if_horizontal:TF` and `\mode_if_inner:TF` because should the user hook contain an unbalanced conditional, things will go off tracks (only these two need changing though, because of the user-level hook inside the conditional body; others should be fine).

Reply:
What kind of  unbalanced conditionals you see could happen here?

Even though they are optimized, it is still a difference between 2 primitive fast scan/jump ahead to reading the code several times as arguments to \use_ii:nnn and the like. (doesn't mean I can't get convinced to use \mode_if_horizontal:TF instead

Reply:
I guess the worry is someone might put `\if...` in the begin hook, `\fi` in the end hook, for excluding blocks of text.

Reply:
yes I see that (for excluding text that would be a pretty weird approach though :-))  but yes you could also have an env added that for some reason starts a conditional, so yes, Phelype has already convinced me

Reply:
Any incomplete TeX conditional command would either throw an error in the wrong place or make the code do something else.  For example, `\AddToHook{para/after}{\fi}` errors with `Extra \else`, and `\AddToHook{para/after}{\else}` makes the kernel hook `\@kernel@after@para@after` be silently ignored (actually makes it only execute if in vmode, followed by the error).

Just to bring some numbers to the table: doing some light benchmarking consistently gives (for 100000 iterations of `\leavevmode\par`) 1.24 s for the raw conditional version and 1.39 s for the higher level version: a 12% relative increase.  Quite a lot.  On the other hand, it's 0.15 s longer for 100000 paragraphs, or 0,0015 ms per paragraph, so not that much in absolute times. Besides, doing the same test with `Lorem ipsum dolor sit amet, consectetur adipisicing elit.` instead of `\leavevmode` raises the times to 2.17 s and 2.34 s, respectively, for the same 100000 paragraphs.

Reply:
Note that using `\if...\fi` like that to ignore a paragraph wouldn't work: the `\fi` would never be seen because it's hidden in the hook and would be skipped by the `\iffalse`.

Reply:
What does save you a couple of cycles in both cases (for the 100000 paragraphs, 0.02 s in the raw conditional case and 0.10 s in the `\mode_if_...:TF` case is to hide the error message with
```latex
\cs_new_protected:Npn \__para_end_vertical_error:
  { \__kernel_msg_error:nnnn {para}{mode}{end}{horizontal} }
```
then use the single token `\__para_end_vertical_error:` in `\para_end:`.

Reply:
> What does save you a couple of cycles in both cases (for the 100000 paragraphs, 0.02 s in the raw conditional case and 0.10 s in the `\mode_if_...:TF` case is to hide the error message with
> 
> ```
> \cs_new_protected:Npn \__para_end_vertical_error:
>   { \__kernel_msg_error:nnnn {para}{mode}{end}{horizontal} }
> ```
> 
> then use the single token `\__para_end_vertical_error:` in `\para_end:`.

that's what I was kind of commenting on in the other thread about error messages, kist passing over them as longish token lists takes time. But this is perhaps something to generally look at  at least in places that are extremely often executed.